broker-core/src/main/java/io/zeebe/broker/subscription/message/state/MessageStateController.java

```diff
@@ -0,0 +1,268 @@
+/*
+ * Zeebe Broker Core
+ * Copyright © 2017 camunda services GmbH (info@camunda.com)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package io.zeebe.broker.subscription.message.state;
+
+import io.zeebe.broker.util.KeyStateController;
+import java.io.File;
+import java.nio.ByteOrder;
+import java.util.ArrayList;
+import java.util.List;
+import org.agrona.DirectBuffer;
+import org.agrona.ExpandableArrayBuffer;
+import org.agrona.concurrent.UnsafeBuffer;
+import org.rocksdb.ColumnFamilyDescriptor;
+import org.rocksdb.ColumnFamilyHandle;
+import org.rocksdb.RocksDB;
+
+public class MessageStateController extends KeyStateController {
+
+  private static final byte[] EXISTENCE = new byte[] {1};
+
+  private static final int TIME_OFFSET = 0;
+  private static final int KEY_OFFSET = TIME_OFFSET + Long.BYTES;
+
+  private static final byte[] MSG_TIME_TO_LIVE_NAME = "msgTimeToLive".getBytes();
+  private static final byte[] MSG_ID_NAME = "messageId".getBytes();
+  private static final byte[] SUB_NAME = "msgSubscription".getBytes();
+  private static final byte[] SUB_SEND_TIME_NAME = "subSendTime".getBytes();
+
+  private final UnsafeBuffer iterateKeyBuffer = new UnsafeBuffer(0, 0);
+
+  protected ColumnFamilyHandle timeToLiveHandle;
+  private ColumnFamilyHandle messageIdHandle;
+
+  private ColumnFamilyHandle subscriptionHandle;
+  private ColumnFamilyHandle subSendTimeHandle;
+
+  private ExpandableArrayBuffer keyBuffer;
+  private ExpandableArrayBuffer valueBuffer;
```

Comment:
Yes that it is why I was a bit unsure to use this, but I have no Idea what to use else. I mean when is the point to shrunk down? Message name, correlation key, id etc are arbitrary :sweat_smile:

Reply:
Well, for the key we could explicitly cap it, I don't see good reasons for users to have message keys of 1MB. As for the value that's a bit harder, so I don't know. We could give some arbitrarily large threshold (e.g. 64MB) where we say if its above this, once we finished reading/writing that value, we resize the buffer back down to 64MB? Although that makes reading the value back a little more complicated since you have to read it in chunks (unless you can query the size of the value before reading?). I guess for now we can leave it as is since I'm not sure what a good strategy is here.

Is there anyway some artificial limit due to our transport layer here or something else in our system? Or could we receive a message value of 8GB for example?

Reply:
> Well, for the key we could explicitly cap it, I don't see good reasons for users to have message keys of 1MB.

I totally aggree, but I don't know if we can easily decide that. We should create an separate issue for this. 

>  We could give some arbitrarily large threshold (e.g. 64MB) where we say if its above this, once we finished reading/writing that value, we resize the buffer back down to 64MB? Although that makes reading the value back a little more complicated since you have to read it in chunks (unless you can query the size of the value before reading?). I guess for now we can leave it as is since I'm not sure what a good strategy is here.

Yeah then you have this problem and repeatedly reallocation's, which are also not cheap.

I think the maximum is given by the dispatcher, which is per default 1 GB and he has 3 partitions so the maximum block size is (1024 * 1024 * 1024) / 3. So it dependents on which dispatcher size is used.

Reply:
https://github.com/zeebe-io/zeebe/issues/1333